src/posit/connect/_api.py

```diff
@@ -0,0 +1,261 @@
+from __future__ import annotations
+
+import posixpath
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Any, Generic, Protocol, TypeVar, cast
+
+from ._json import Jsonifiable, JsonifiableDict, ResponseAttrs
+
+if TYPE_CHECKING:
+    from .context import Context
+
+# Design Notes:
+# * Perform API calls on property retrieval
+# * Dictionary endpoints: Retrieve all attributes during init unless provided
+# * List endpoints: Do not retrieve until `.fetch()` is called directly. Avoids cache invalidation issues.
+# * Only expose methods needed for `ReadOnlyDict`.
+#   * Ex: When inheriting from `dict`, we'd need to shut down `update`, `pop`, etc.
+# * Use `ApiContextProtocol` to ensure that the class has the necessary attributes for API calls.
+#    * Inherit from `EndpointMixin` to add all helper methods for API calls.
+# * Classes should write the `path` only once within its init method.
+#    * Through regular interactions, the path should only be written once.
+
+
+class ApiContextProtocol(Protocol):
+    _ctx: Context
+    _path: str
+
+
+# TODO-future?; Add type hints for the ReadOnlyDict class
+# ArgsT = TypeVar("ArgsT", bound="ResponseAttrs")
+
+
+class ReadOnlyDict(
```

Comment:
For what it's worth, I think this is fairly similar to the original functionality implemented and then removed in this commit: https://github.com/posit-dev/posit-sdk-py/commit/50e78ac5b8d608c34144553db9727203045a04bb

Locking the __setitem__ post initialization is interesting. Did you explore using `frozendict`? https://pypi.org/project/frozendict/

It's worth considering if __setitem__ should make a PATCH request to have the server reflect the local object state. Based on our current conversations, I am leaning towards no. Instead, a call to `update` should be used. 

Hm... If we modified `Resource` to inherit from `frozendict` instead of `dict`, then a lot of the `update` behavior is simplified.

I pulled on the thread a little bit... Here is a modification to `Resource` to use `frozendict`. It may be odd to look at, but I think it's worth exploring further.

```py
class Resource(frozendict):
    params: ResourceParameters

    def __new__(cls: Type["Resource"], params: ResourceParameters, *args, **kwargs):
        instance = super(Resource, cls).__new__(cls, *args, **kwargs)
        instance.params = params
        return instance

    def __init__(self, params: ResourceParameters, *args, **kwargs):
        # Skip setting attributes in __init__ to maintain immutability
        super().__init__(*args, **kwargs)

    def __getattr__(self, name):
        if name in self:
            warnings.warn(
                f"Accessing the field '{name}' via attribute is deprecated and will be removed in v1.0.0. "
                f"Please use __getitem__ (e.g., {self.__class__.__name__.lower()}['{name}']) for field access instead.",
                DeprecationWarning,
                stacklevel=2,
            )
            return self[name]
        return None
        ```

Reply:
I have not looked through `frozendict` before. Neat project!
Nice trick with `__new__`. Clever.

-----------------------------

Couple ideas I was trying to go for in `ApiDictEndpoint` which inherits from `EndpointMixin` and this `ReadOnlyDict`.

1. Do not expose `.params`. It is internal information. (Redistributed to `._ctx` and `._path`)
2. Do not expose any methods through implementation coincidence. Ex: `frozendict` is awesome but exposes methods like `.setdefault(key, default)`, which I do not want to show to users.

Maybe a [`collections.abc.Mapping`](https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping) would work? It would only exposes `keys`, `items`, `values`, and `get` (plus `__getitem__`, `__iter__`, `__len__`, `__contains__`, `__eq__`, and `__ne__`).

Reply:
Using a read only dictionary makes it quite clear!

Reply:
Talked through w/ @tdstein in person and am using `Mapping` with minimal method exposure as our way forward.